### ISSUE TITLE
2018-08-11 edge + the-state-of-art

### ISSUE THREAD

**Issue by yzbx** [ISCAS007/torchseg#9]
# miou
| Network | backbone | note | voc | cityscapes | ade20k |
| - | :-: | - | - | - | - |
| deeplabv3+ | resnet50 | - | 64-74 | - | - |
| deeplabv3+ | resnet101 | - | 68-75 | - | - |
| deeplabv3+ | resnet101 | OS=16 | 76.66 | 77.23 | - |
| deeplabv3+ | resnet101 | OS=8 | 78.05 | 77.82 | - |
| deeplabv3+ | resnet101 | OS=8,MS,Flip | 79.35 | 79.30 | - |
| deeplabv3+ | resnet101 | coarse | - | 81.3 | - |
| deeplabv3+ | resnet101 | crop size=321 | 67.22 | - | - |
| deeplabv3+ | resnet101 | crop size=513 | 77.21 | - | - |
| deeplabv3+ | resnet101 | batch size=4 | 64.43 | - | - |
| deeplabv3+ | resnet101 | batch size=8 | 75.76 | - | - |
| deeplabv3+ | resnet101 | batch size=12 | 76.49 | - | - |
| deeplabv3+ | resnet101 | batch size=16 | 77.21 | - | - |
| deeplabv3+ | resnet101 | duplicate hard classes | 85.7 | - | - |

| Network | backbone | note | voc | cityscapes | ade20k |
| - | :-: | - | - | - | - |
| pspnet | resnet50 | - | - | 
| pspnet | resnet101 | - | - | - | 43.39 |
| pspnet | resnet152 | - | - | 

multi-scale {0.5,0.8,1.0,1.2,1.5,2.0} testing will help denseASPP miou to 80.6%
|Network|backbone|cityscapes|
| - | - | - |
|denseASPP|densenet121|76.2|
|denseASPP|densenet169|77.7|
|denseASPP|densenet201|78.9|

**Comment by yzbx** [comment]
https://www.slideshare.net/mitmul/unofficial-pyramid-scene-parsing-network-cvpr-2017

**Comment by yzbx** [comment]
# add regulator l1 and l2 loss
```
python test/pspnet_test.py --batch_size=4 --optimizer=sgd --learning_rate=0.01 --midnet_name=aspp --backbone_pretrained=True --note='reg'
```

**Comment by yzbx** [comment]
# tensorflow
- https://github.com/GeorgeSeif/Semantic-Segmentation-Suite
basic implement of lots of models
- https://github.com/ISCAS007/PSPNet-tensorflow
caffe2tensorflow

|dataset|accuracy|
|-|-|
|cityscapes|77%|
|ade20k|40%|
- https://github.com/holyseven/PSPNet-TF-Reproduce
    - L2-SP regularization
    - sync batch norm
    - 80.3% without post-processing methods on cityscapes test set
    
# pytorch
- https://github.com/CSAILVision/semantic-segmentation-pytorch
    - UPerNet without dilated convolution, but comparable or even better compared with PSPNet

**Comment by yzbx** [comment]
# abnormal object detection
https://github.com/ISCAS007/keras-yolo3
1. download and convert weights
2. demo video and target area, target object detection

**Comment by yzbx** [comment]
# tensorflow example
- [tfrecord](https://github.com/Hvass-Labs/TensorFlow-Tutorials/blob/master/18_TFRecords_Dataset_API.ipynb)
- [multigpu basic](https://github.com/aymericdamien/TensorFlow-Examples/blob/master/notebooks/6_MultiGPU/multigpu_basics.ipynb)
- [multigpu cnn](https://github.com/aymericdamien/TensorFlow-Examples/blob/master/notebooks/6_MultiGPU/multigpu_cnn.ipynb)

**Comment by yzbx** [comment]
# problem
we need use edge, convert the label from tfrecord or from tensor
>if use tensor
1. we need use opencv lib to do edge detection and dilate
2. opencv only support numpy array
3. need convert data from tensor to numpy array
4. need use tf.Session() to accomplish the convert
5. cannot use default session in tensorflow, no interface found for tf.Session()
>if use tfrecord
1. not support dynamic edge_width

# ways
1. [convert tensorflow model and weight to pytorch](https://github.com/Microsoft/MMdnn)
2. remove distrubute support, supervisitor support.

**Comment by yzbx** [comment]
# problem
1. use official tensorflow preprocessing
2. use feed_dict
3. cannot convert a tensor to tensor

```
        for epoch in range(epoches):
            for i, (images, labels, edges) in enumerate(data_loader):
                tf_images_4d,tf_labels_4d=batch_preprocess_image_and_label(images.numpy(),labels.numpy(),FLAGS,ignore_label,is_training=True)
#                tf_labels_4d = tf.expand_dims(tf_labels_3d, axis=-1)
                
                print(tf_images_4d.shape,tf_labels_4d)
                sess.run(fetches=[optimizer.minimize(total_loss), total_loss], feed_dict={
                         images: tf_images_4d, labels: tf_labels_4d})
```

# ways
1. use numpy preprocessing
2. find a way to feed with tensorflow tensor

**Comment by yzbx** [comment]
# todo
1. FailedPreconditionError (see above for traceback): Attempting to use uninitialized value logits/semantic/biases/Momentum
2. arbitrary input size
3. edge preprocess
4. voc, ade20k dataset suport
